diff --git a/chunkie/+chnk/+rcip/level.m b/chunkie/+chnk/+rcip/level.m
index 1fa13e0..860fd54 100644
--- a/chunkie/+chnk/+rcip/level.m
+++ b/chunkie/+chnk/+rcip/level.m
@@ -16,7 +16,7 @@ classdef level
     end
 
     methods
-        function obj = level(r,d,h,adj,p,pw,glw)
+        function obj = level(r,d,h,adj,p,pw,glw,starfun)
             obj.r = r;
             obj.d = d;
             obj.h = h;
@@ -25,6 +25,7 @@ classdef level
             obj.pw = pw;
             obj.glw = glw;
             obj.star = false(size(r,3));
+            obj.starfun = starfun;
         end
 
         function obj = buildstar(obj)
@@ -152,7 +153,12 @@ classdef level
             
         end
 
+        function kcirc(obj,kstar)
+            % compute the kcirc matrix at the current level
 
+
+
+        end
         % function kstar_compressed = compress_k_star(obj,kstar)
         %     P = obj.p;
         %     PWT = obj.pw';
@@ -186,3 +192,9 @@ classdef level
         % end
     end
 end 
+
+
+function mat = build_mat_kcirc()
+
+
+end
\ No newline at end of file
diff --git a/chunkie/chunkermat.m b/chunkie/chunkermat.m
index 14f346c..eaaa468 100644
--- a/chunkie/chunkermat.m
+++ b/chunkie/chunkermat.m
@@ -321,7 +321,9 @@ for i=1:nchunkers
 
         % mark off the near and self interactions
         for ich = 1:chnkr.nch
+            
             for jch = [ich,chnkr.adj(1,ich),chnkr.adj(2,ich)]
+                if jch < 0, continue, end
                 flag((jch - 1)*chnkr.k+(1:chnkr.k), ich) = 0;
             end
         end
diff --git a/devtools/test/rcipTest.m b/devtools/test/rcipTest.m
index c517f37..f899e4b 100644
--- a/devtools/test/rcipTest.m
+++ b/devtools/test/rcipTest.m
@@ -5,7 +5,7 @@
 % problem on a domain defined by two arcs of circles meeting at two vertices
 
 clearvars; close all;
-addpaths_loc();
+% addpaths_loc();
 
 ncurve = 2;
 chnkr(1,ncurve) = chunker();
diff --git a/devtools/test/rcip_analyticTest.m b/devtools/test/rcip_analyticTest.m
index 06fa0ad..f85f9f6 100644
--- a/devtools/test/rcip_analyticTest.m
+++ b/devtools/test/rcip_analyticTest.m
@@ -5,7 +5,7 @@
 % problem on a domain defined by two arcs of circles meeting at two vertices
 
 clearvars; close all;
-addpaths_loc();
+% addpaths_loc();
 
 ncurve = 2;
 chnkr(1,ncurve) = chunker();
@@ -13,16 +13,13 @@ chnkr(1,ncurve) = chunker();
 % set wave number
 zk = 1.1;
 
-
-
-
 nch = 5*ones(1,ncurve);
 
 a = -1.0;
 b = 1.0;
 
 % set angles for top and bottom curve
-theta = [pi/2.2 pi/2.4];
+theta = [0.1 0.1];
 
 % set parametrization start and end points for chunkie objects
 tab = [-theta(1)/2 -theta(2)/2; theta(1)/2 theta(2)/2];
@@ -85,7 +82,7 @@ strengths = randn(ns,1);
 nt = 20;
 ts = 0.0+2*pi*rand(1,nt);
 targets = [cos(ts);sin(ts)];
-targets = 0.2*targets;
+targets = 0.002*targets;
 
 scatter(sources(1,:),sources(2,:),'o');
 scatter(targets(1,:),targets(2,:),'x');
@@ -97,7 +94,7 @@ axis equal
 chnkrtotal = merge(chnkr);
 fkern = @(s,t) -2*chnk.helm2d.kern(zk,s,t,'D');
 np = chnkrtotal.k*chnkrtotal.nch;
-start = tic; D = chunkermat(chnkrtotal,fkern);
+start = tic; D = chunkermat(chnkrtotal,fkern,struct('adaptive_correction',true));
 t1 = toc(start);
 
 
@@ -150,6 +147,8 @@ ndim = 1;
 nsub = 100;
 
 opts = [];
+sysmat_pre = sysmat;
+pre = eye(np);
 
 for icorner=1:ncorner
     clist = corners{icorner}.clist;
@@ -182,10 +181,14 @@ for icorner=1:ncorner
     toc
 
     sysmat(starind,starind) = inv(R{icorner}) - eye(2*ngl*nedge*ndim);
+    sysmat_pre(starind,starind) = 0;
+    pre(starind,starind) = R{icorner};
 end
 sysmat = sysmat + eye(np);
+sysmat_pre = eye(np) + sysmat_pre * pre;
 
 sol = gmres(sysmat,ubdry,np,eps*20,np);
+solp = gmres(sysmat_pre,ubdry,np,eps*20,np);
 
 opts.usesmooth=true;
 opts.verb=false;
